File: python_workflow/tests/test_validation.py
import inspect
import math

import numpy as np
import pandas as pd

from lexsync.validation import (
    balance_check,
    cohens_d,
    cohens_d_ci,
    describe_stimuli,
    match_report,
    match_report_continuous,
    tost_equiv,
    variance_ratio,
)


def _cont_stim():
    return pd.DataFrame({
        "word": ["a", "b", "c", "d"], "condition": "continuous",
        "frequency": [2.0, 3.0, 4.0, 5.0], "length": [3, 4, 3, 4],
        "n_density": [1, 2, 1, 2], "old20": [1.5, 1.6, 1.5, 1.6],
    })


def test_match_report_continuous_shape(schema):
    rep = match_report_continuous(_cont_stim(), "frequency",
                                  ["length", "n_density", "old20"], schema)
    assert set(rep) == {"descriptives", "comparisons"}
    c = rep["comparisons"]
    assert list(c.columns) == ["dimension", "role", "pearson_r", "predictor_span"]
    assert list(c["role"]) == ["predictor", "control", "control", "control"]
    assert c.loc[c.dimension == "frequency", "predictor_span"].iloc[0] == 3.0


def test_cohens_d():
    assert cohens_d([1, 2, 3, 4], [1, 2, 3, 4]) == 0
    assert abs(cohens_d([5, 6, 7, 8], [1, 2, 3, 4])) > 1


def test_tost_equivalent_for_matched_samples():
    x = [4, 5, 6] * 20
    y = [4, 5, 6] * 20
    assert tost_equiv(x, y, bound_d=0.5)["equivalent"] is True


def test_tost_default_bound_is_the_schema_bound():
    # The bound is the schema's smallest effect size of interest (Lakens, 2017);
    # pinned here because every other test passes bound_d explicitly. The p-value is
    # the same literal asserted in test-validation.R, so the engines cannot drift apart.
    assert inspect.signature(tost_equiv).parameters["bound_d"].default == 0.5
    assert round(tost_equiv(range(10, 20), range(11, 21))["p"], 9) == 0.354383811


def test_cohens_d_ci_brackets_point_estimate():
    x = [5, 6, 7, 8] * 10
    y = [1, 2, 3, 4] * 10
    ci = cohens_d_ci(x, y)
    assert ci["ci_low"] <= ci["d"] <= ci["ci_high"]
    # The point estimate matches the standalone Cohen's d.
    assert abs(ci["d"] - cohens_d(x, y)) < 1e-9


def test_cohens_d_ci_width_shrinks_with_more_items():
    # The interval's dependence on the number of items is the property that makes
    # it a robust complement to the binary TOST verdict: few items -> wide CI.
    small = cohens_d_ci([4, 5, 6] * 2, [4, 5, 7] * 2)
    large = cohens_d_ci([4, 5, 6] * 40, [4, 5, 7] * 40)
    assert (small["ci_high"] - small["ci_low"]) > (large["ci_high"] - large["ci_low"])


def test_cohens_d_ci_coheres_with_tost():
    # The 90% CI lying inside +/- bound is equivalent to a TOST pass at .05.
    x = [4, 5, 6] * 20
    y = [4, 5, 6] * 20
    ci = cohens_d_ci(x, y, alpha=0.05)
    within = (ci["ci_low"] > -0.5) and (ci["ci_high"] < 0.5)
    assert within == bool(tost_equiv(x, y, bound_d=0.5, alpha=0.05)["equivalent"])


def test_cohens_d_ci_constant_dimension_is_a_point():
    # A dimension fixed by the pool (e.g. two-character words) has no uncertainty.
    ci = cohens_d_ci([2, 2, 2, 2], [2, 2, 2, 2])
    assert ci == {"d": 0.0, "ci_low": 0.0, "ci_high": 0.0}


def test_zero_pooled_sd_with_unequal_means_is_undefined_not_zero():
    # Two conditions each constant at a different value differ by infinitely many
    # SDs: reporting d = 0 (perfect balance) inverted the truth. The equal-constant
    # case is pinned by the zh_freqcontrast golden and must stay exactly zero.
    x = [2, 2, 2, 2]
    y = [3, 3, 3, 3]
    assert cohens_d(x, y) is None
    assert cohens_d_ci(x, y) == {"d": None, "ci_low": None, "ci_high": None}
    assert tost_equiv(x, y)["equivalent"] is False  # already correct on this branch
    assert variance_ratio(y, x) == 1.0              # unchanged: both spreads are zero
    assert cohens_d(x, x) == 0                      # equal constants stay exactly zero


def test_match_report_carries_an_undefined_d_as_a_missing_cell(schema):
    # Serialised as an empty CSV cell and a null JSON value, exactly like the other
    # missing statistics, and identically in both engines.
    stim = pd.DataFrame({"word": ["aa", "bb", "cc", "dd"],
                         "condition": ["a", "a", "b", "b"],
                         "length": [2, 2, 3, 3]})
    cmp = match_report(stim, ["length"], schema)["comparisons"]
    assert cmp["cohens_d"].isna().iloc[0]
    assert cmp["d_ci_low"].isna().iloc[0]
    assert cmp["d_ci_high"].isna().iloc[0]
    assert cmp["equivalent"].iloc[0] == False  # noqa: E712 -- numpy bool, not None
    assert cmp["var_ratio"].iloc[0] == 1.0


def test_variance_ratio():
    assert abs(variance_ratio([1, 2, 3, 4], [1, 2, 3, 4]) - 1.0) < 1e-9  # equal spread
    assert variance_ratio([0, 5, 10, 15], [4, 5, 6, 7]) > 1              # condition wider
    assert variance_ratio([4, 5, 6, 7], [0, 5, 10, 15]) < 1             # condition narrower
    assert variance_ratio([1, 2, 3], [5, 5, 5]) is None                  # constant reference
    assert variance_ratio([5, 5, 5], [2, 2, 2]) == 1.0                   # both constant


def test_describe_and_balance():
    df = pd.DataFrame({"condition": ["a", "a", "b", "b"], "x": [1, 2, 3, 4]})
    d = describe_stimuli(df, ["x"])
    assert {"group", "dimension", "n", "mean", "sd"} <= set(d.columns)
    assert len(d) == 2
    assert len(balance_check(pd.DataFrame({"condition": ["a", "a", "a", "b"]}), "condition")) == 1
    assert len(balance_check(pd.DataFrame({"condition": ["a", "a", "b", "b"]}), "condition")) == 0


def test_describe_stimuli_groups_in_first_appearance_order():
    # Not sorted order: this is the order validation.R's appearance-ordered factor
    # yields, and the order in which match_report() takes its anchor.
    df = pd.DataFrame({"condition": ["b", "b", "a", "a"], "x": [1, 2, 3, 4]})
    assert list(describe_stimuli(df, ["x"])["group"]) == ["b", "a"]


def test_describe_stimuli_all_na_dimension_is_missing_not_infinite():
    df = pd.DataFrame({"condition": ["a", "a"], "x": [np.nan, np.nan]})
    d = describe_stimuli(df, ["x"])
    assert d["n"].iloc[0] == 0
    assert math.isnan(d["min"].iloc[0])
    assert math.isnan(d["max"].iloc[0])
    assert not math.isinf(d["min"].iloc[0])
    assert not math.isinf(d["max"].iloc[0])

File: R_workflow/tests/testthat/test-validation.R
test_that("cohens_d is zero for identical samples and large for separated ones", {
  expect_equal(cohens_d(c(1, 2, 3, 4), c(1, 2, 3, 4)), 0)
  expect_gt(abs(cohens_d(c(5, 6, 7, 8), c(1, 2, 3, 4))), 1)
})

test_that("tost_equiv flags closely matched samples as equivalent", {
  set.seed(1)
  x <- rep(c(4, 5, 6), 20)
  y <- rep(c(4, 5, 6), 20)
  tt <- tost_equiv(x, y, bound_d = 0.5)
  expect_true(isTRUE(tt$equivalent))
})

test_that("tost_equiv defaults to the schema's 0.5 bound", {
  # The bound is the schema's smallest effect size of interest (Lakens, 2017);
  # pinned here because every other test passes bound_d explicitly. The p-value is
  # the same literal asserted in test_validation.py, so the engines cannot drift apart.
  expect_identical(formals(tost_equiv)$bound_d, 0.5)
  expect_equal(round(tost_equiv(10:19, 11:20)$p, 9), 0.354383811)
})

test_that("cohens_d_ci brackets the point estimate and matches cohens_d", {
  x <- rep(c(5, 6, 7, 8), 10); y <- rep(c(1, 2, 3, 4), 10)
  ci <- cohens_d_ci(x, y)
  expect_lte(ci$ci_low, ci$d)
  expect_lte(ci$d, ci$ci_high)
  expect_equal(ci$d, cohens_d(x, y), tolerance = 1e-9)
})

test_that("cohens_d_ci width shrinks as the number of items grows", {
  small <- cohens_d_ci(rep(c(4, 5, 6), 2),  rep(c(4, 5, 7), 2))
  large <- cohens_d_ci(rep(c(4, 5, 6), 40), rep(c(4, 5, 7), 40))
  expect_gt(small$ci_high - small$ci_low, large$ci_high - large$ci_low)
})

test_that("cohens_d_ci within the bound coheres with the TOST verdict", {
  x <- rep(c(4, 5, 6), 20); y <- rep(c(4, 5, 6), 20)
  ci <- cohens_d_ci(x, y, alpha = 0.05)
  within <- ci$ci_low > -0.5 && ci$ci_high < 0.5
  expect_identical(within, isTRUE(tost_equiv(x, y, bound_d = 0.5, alpha = 0.05)$equivalent))
})

test_that("cohens_d_ci is a point for a constant dimension", {
  ci <- cohens_d_ci(c(2, 2, 2, 2), c(2, 2, 2, 2))
  expect_equal(c(ci$d, ci$ci_low, ci$ci_high), c(0, 0, 0))
})

test_that("a zero pooled SD with unequal means is undefined, not zero", {
  # Two conditions each constant at a different value differ by infinitely many
  # SDs: reporting d = 0 (perfect balance) inverted the truth. The equal-constant
  # case is pinned by the zh_freqcontrast golden and must stay exactly zero.
  x <- c(2, 2, 2, 2); y <- c(3, 3, 3, 3)
  expect_true(is.na(cohens_d(x, y)))
  ci <- cohens_d_ci(x, y)
  expect_true(all(is.na(c(ci$d, ci$ci_low, ci$ci_high))))
  expect_false(tost_equiv(x, y)$equivalent)  # already correct on this branch
  expect_equal(variance_ratio(y, x), 1)      # unchanged: both spreads are zero
  expect_equal(cohens_d(x, x), 0)            # equal constants stay exactly zero
})

test_that("match_report carries an undefined d as a missing cell", {
  # Serialised as an empty CSV cell and a null JSON value, exactly like the other
  # missing statistics, and identically in both engines.
  schema <- yaml::read_yaml(system.file("extdata", "schema.yaml", package = "lexsync"))
  stim <- data.frame(word = c("aa", "bb", "cc", "dd"),
                     condition = c("a", "a", "b", "b"),
                     length = c(2, 2, 3, 3), stringsAsFactors = FALSE)
  cmp <- match_report(stim, "length", schema)$comparisons
  expect_true(is.na(cmp$cohens_d))
  expect_true(is.na(cmp$d_ci_low))
  expect_true(is.na(cmp$d_ci_high))
  expect_false(cmp$equivalent)
  expect_equal(cmp$var_ratio, 1)
})

test_that("match_report gives a single-condition design an empty comparisons frame", {
  # There is nothing to compare against the anchor, so the frame comes back with its
  # columns and no rows. rbind() over an empty list used to return NULL, which
  # run_pipeline's seq_len(nrow(...)) loop died on with "argument must be coercible
  # to non-negative integer", naming neither the design nor the cause, while the
  # Python engine carried on and wrote a comparisons CSV with no header at all.
  # Pinned identically in tests/test_validation.py.
  schema <- yaml::read_yaml(system.file("extdata", "schema.yaml", package = "lexsync"))
  stim <- data.frame(word = c("aa", "bb"), condition = c("only", "only"),
                     length = c(2, 3), stringsAsFactors = FALSE)
  cmp <- match_report(stim, "length", schema)$comparisons
  expect_s3_class(cmp, "data.frame")
  expect_equal(nrow(cmp), 0L)
  expect_identical(names(cmp),
                   c("condition", "reference", "dimension", "cohens_d", "d_ci_low",
                     "d_ci_high", "var_ratio", "tost_p", "equivalent"))
  # The loop that used to die on NULL now runs zero times.
  expect_length(seq_len(nrow(cmp)), 0L)
})

test_that("variance_ratio flags unequal spread", {
  expect_equal(variance_ratio(c(1, 2, 3, 4), c(1, 2, 3, 4)), 1)       # equal spread
  expect_gt(variance_ratio(c(0, 5, 10, 15), c(4, 5, 6, 7)), 1)        # condition wider
  expect_lt(variance_ratio(c(4, 5, 6, 7), c(0, 5, 10, 15)), 1)        # condition narrower
  expect_true(is.na(variance_ratio(c(1, 2, 3), c(5, 5, 5))))          # constant reference
  expect_equal(variance_ratio(c(5, 5, 5), c(2, 2, 2)), 1)             # both constant
})

test_that("match_report_continuous returns predictor and control rows", {
  schema <- yaml::read_yaml(system.file("extdata", "schema.yaml", package = "lexsync"))
  stim <- data.frame(word = c("a", "b", "c", "d"), condition = "continuous",
                     frequency = c(2, 3, 4, 5), length = c(3, 4, 3, 4),
                     n_density = c(1, 2, 1, 2), old20 = c(1.5, 1.6, 1.5, 1.6),
                     stringsAsFactors = FALSE)
  rep <- match_report_continuous(stim, "frequency", c("length", "n_density", "old20"), schema)
  expect_true(all(c("descriptives", "comparisons") %in% names(rep)))
  cmp <- rep$comparisons
  expect_identical(names(cmp), c("dimension", "role", "pearson_r", "predictor_span"))
  expect_identical(cmp$role, c("predictor", "control", "control", "control"))
  expect_equal(cmp$predictor_span[1], 3)
})

test_that("describe_stimuli summarises per group", {
  df <- data.frame(condition = c("a", "a", "b", "b"), x = c(1, 2, 3, 4), stringsAsFactors = FALSE)
  d <- describe_stimuli(df, "x")
  expect_true(all(c("group", "dimension", "n", "mean", "sd") %in% names(d)))
  expect_equal(nrow(d), 2)
})

test_that("describe_stimuli groups in order of first appearance", {
  # Not locale-collated order: this is the order validation.py's groupby(sort = FALSE)
  # yields, and the order in which match_report() takes its anchor from unique().
  df <- data.frame(condition = c("b", "b", "a", "a"), x = c(1, 2, 3, 4), stringsAsFactors = FALSE)
  expect_identical(describe_stimuli(df, "x")$group, c("b", "a"))
})

test_that("describe_stimuli reports an all-NA dimension as missing, not infinite", {
  df <- data.frame(condition = c("a", "a"), x = c(NA_real_, NA_real_))
  d <- describe_stimuli(df, "x")
  expect_identical(d$n, 0L)
  expect_true(is.na(d$min))
  expect_true(is.na(d$max))
  expect_false(is.infinite(d$min))
  expect_false(is.infinite(d$max))
})

test_that("balance_check detects imbalance", {
  balanced <- data.frame(condition = c("a", "a", "b", "b"))
  unbalanced <- data.frame(condition = c("a", "a", "a", "b"))
  expect_length(balance_check(balanced, "condition"), 0)
  expect_length(balance_check(unbalanced, "condition"), 1)
})

# The message reaches the run log, so the two engines must word it alike. Byte
# order is the one ordering neither table() nor pandas' value_counts reaches on
# its own; test_validation.py pins the same sentence.
test_that("balance_check names the levels in byte order", {
  df <- data.frame(condition = c("low", "high", "high", "high", "mid", "mid"),
                   stringsAsFactors = FALSE)
  expect_identical(balance_check(df, "condition"),
                   "Column 'condition' is unbalanced: high=3, low=1, mid=2")
})

named_stim <- function() {
  data.frame(word = letters[1:4], condition = c("hi", "hi", "lo", "lo"),
             frequency = c(1, 2, 3, 4), stringsAsFactors = FALSE)
}

test_that("a dimension the stimuli do not have is refused", {
  # A misspelt dimension used to give a full table of NAs, with a Cohen's d of zero
  # beside it, and a bare KeyError in the Python engine. Mirrored in
  # test_validation.py.
  stim <- named_stim()
  msg <- "lexsync: cannot report on dimension(s) the stimuli do not have: 'frequncy'."
  expect_error(describe_stimuli(stim, "frequncy"), msg, fixed = TRUE)
  expect_error(match_report(stim, "frequncy", list()), msg, fixed = TRUE)
  expect_error(match_report_continuous(stim, "frequncy", character(0), list()),
               msg, fixed = TRUE)
  # Offenders are named in byte order, as balance_lists names them.
  expect_error(match_report(stim, c("zz", "aa"), list()),
               "do not have: 'aa', 'zz'.", fixed = TRUE)
})

test_that("a grouping column the stimuli do not have is refused", {
  stim <- named_stim()
  expect_error(describe_stimuli(stim, "frequency", by = "cond"),
               "lexsync: cannot group the stimuli by column 'cond', which they do not have.",
               fixed = TRUE)
  names(stim)[2] <- "cond"
  msg <- "lexsync: cannot group the stimuli by column 'condition', which they do not have."
  expect_error(match_report(stim, "frequency", list()), msg, fixed = TRUE)
  expect_error(match_report_continuous(stim, "frequency", character(0), list()),
               msg, fixed = TRUE)
})
